Dev 3.20

GigaZiv opened the pull request:
Выполнено задание 20 спринта

С уважением Жуков Игорь

GigaZiv committed:
Dev 3.20
Выполнено задание 20 спринта

С уважением Жуков Игорь

ArturNurtdinov commented on app/src/main/java/rs/example/playlistmaker/player/ui/viewmodel/PlayerViewModel.kt:
```diff
@@ -18,22 +21,24 @@ class PlayerViewModel(private val playerInteractor: PlayControl) :
+    private var timerJob: Job? = null
+
+    private fun startTimer(state: PlayerState) {
+        timerJob = viewModelScope.launch(Dispatchers.Default) {
```
Перед запуском новой корутины на всякий случай лучше отменить старую, чтобы гарантировать, что только 1 цикл будет обновлять время трека:
 ```
 timerJob?.cancel()
 timerJob = ...
 ```

ArturNurtdinov commented on app/src/main/java/rs/example/playlistmaker/player/ui/viewmodel/PlayerViewModel.kt:
```diff
@@ -18,22 +21,24 @@ class PlayerViewModel(private val playerInteractor: PlayControl) :
+
+    private fun startTimer(state: PlayerState) {
+        timerJob = viewModelScope.launch(Dispatchers.Default) {
+            while (state == PlayerState.PLAYING) {
```
Тут же стейт не будет меняться, поэтому по сути цикл while либо отработает 0 итераций, либо будет бесконечным, потому что state захватывается в лямбду при вызове функции `startTimer`

ArturNurtdinov commented on app/src/main/java/rs/example/playlistmaker/player/ui/viewmodel/PlayerViewModel.kt:
```diff
@@ -44,17 +49,19 @@ class PlayerViewModel(private val playerInteractor: PlayControl) :
-            mainThreadHandler.removeCallbacks(
-                progressTimeRunnable
-            )
+        if (state == PlayerState.PLAYING) startTimer(state) else cancelTimer()
```
Вот тут уже стейт берётся актуальный из интерактора, поэтому тут корректно будет отрабатывать, поэтому рекомендую также в функции `startTimer` внутри цикла `while` ориентироваться на актуальный стейт, а не переданный в качестве аргумента

ArturNurtdinov commented on app/src/main/java/rs/example/playlistmaker/search/data/network/RetrofitNetworkClient.kt:
```diff
@@ -11,17 +13,21 @@ class RetrofitNetworkClient(
-        return body?.apply { resultCode = resp.code() } ?: Response().apply {
-            resultCode = resp.code()
+
+        return withContext(Dispatchers.IO) {
```
На IO лучше не переключать самостоятельно, т.к. OkHttp под капотом использует свой пул потоков и самостоятельно переключает поток при отправке запроса в сеть и для ожидания ответа

GigaZiv committed:
Dev 3.20.1

Исправлена константа 20 спринта

«Аудиоплеер» используются корутины, прогресс воспроизведения обновляется раз в 300 миллисекунд

С уважением Жуков Игорь
